### PR TITLE
fix(1030): [1] Add method to call generating build JWT API and replace temporal JWT

### PR DESCRIPTION
**Pull request by sakka2**
## Context
This PR is related to these issue before.
See this [issue #998](https://github.com/screwdriver-cd/screwdriver/issues/998).
See this [issue #1075](https://github.com/screwdriver-cd/screwdriver/issues/1075).

## Objective 
- Inside executor, JWT for build will be generated using temporal JWT which is passed in from model.
- All executors extends this `executor-base`, so add `exchangeTokenForBuild()` method here.
- This method will get argument `config` and `buildTimeout`. `config` object contains `token` which is temporal JWT. And `buildTimeout` will be passed when executor is `k8s` or `k8s-vm`. Other executor, buildTimeout will be default value 90 minutes.
- This method calls [API which generate build JWT](https://github.com/screwdriver-cd/screwdriver/pull/1087), with temporal JWT and buildTimeout value. 
- With the API response (It will be the build JWT), finally the `config.token` will be replaced to the build JWT.

## References
- Related: (other PRs)
　https://github.com/screwdriver-cd/screwdriver/pull/1087
　https://github.com/screwdriver-cd/models/pull/245
　https://github.com/screwdriver-cd/executor-base/pull/42
　https://github.com/screwdriver-cd/executor-k8s/pull/84
　https://github.com/screwdriver-cd/executor-k8s-vm/pull/33
　https://github.com/screwdriver-cd/executor-docker/pull/25
　https://github.com/screwdriver-cd/executor-jenkins/pull/20

## TODO
- [x] Add unit test

### CODE REVIEW

File: test/index.test.js [hunk 1]
```diff
@@ -151,4 +157,68 @@ describe('index test', () => {
...
+            const returnMessage = `Error: Invalid buildTimeout value: ${buildTimeout}`;
+
+            await instance.exchangeTokenForBuild(postConfig, buildTimeout).then(() => {
+                throw new Error('did not fail');
+            }).catch((err) => {
+                assert.equal(err.message, returnMessage);
+            });
+        });
+
+        it('It returns error if response code is not 200', async () => {
```
Comment (tk3fftk) [inline test/index.test.js]: `It` is duplicated.

File: index.js [hunk 2]
```diff
@@ -121,6 +123,41 @@ class Executor {
...
+                strictSSL: false,
+                json: true
+            };
+
+            const response = await request(options);
+
+            if (response.statusCode !== 200) {
+                throw new Error(`Failed to exchange build token: ${JSON.stringify(response.body)}`);
+            }
+            config.token = response.body.token;
```
Comment (tk3fftk) [inline index.js]: I think this replace only works in this function's scope.

File: index.js [hunk 3]
```diff
@@ -121,6 +123,41 @@ class Executor {
...
+        } catch (err) {
+            throw new Error(err);
```
Comment (tk3fftk) [inline index.js]: I don't know we need this catch or not...

File: test/index.test.js (hunk 1 above)
Comment (sakka2) [reply test/index.test.js]: Thanks, I deleted.

File: index.js (hunk 3 above)
Comment (sakka2) [reply index.js]: I deleted try/catch.

File: index.js [hunk 4]
```diff
@@ -121,6 +123,38 @@ class Executor {
...
     }
+
+    /**
+     * Get build JWT with temporal JWT using API and replace config.token
+     * @method exchangeTokenForBuild
+     * @param  {Object}  config          A configuration object
+     * @param  {String}  config.apiUrl   Base URL for Screwdriver API
+     * @param  {String}  config.buildId  Build ID
+     * @param  {String}  config.token    Temporal JWT
+     * @param  {String}  buildTimeout    Build timeout value which will be JWT expires time
```
Comment (tk3fftk) [inline index.js]: Do we need a JSDoc of the return value?

Comment (sakka2) [reply index.js]: Thanks, I added the return value.

File: index.js [hunk 5]
```diff
@@ -125,13 +125,14 @@ class Executor {
...
     /**
-     * Get build JWT with temporal JWT using API and replace config.token
+     * Get JWT for build by using temporal JWT via API
      * @method exchangeTokenForBuild
      * @param  {Object}  config          A configuration object
      * @param  {String}  config.apiUrl   Base URL for Screwdriver API
      * @param  {String}  config.buildId  Build ID
      * @param  {String}  config.token    Temporal JWT
      * @param  {String}  buildTimeout    Build timeout value which will be JWT expires time
+     * @return {String}  token           JWT for build
```
Comment (tk3fftk) [inline index.js]: This is async function so I think it returns `Promise`.

File: index.js [hunk 6]
```diff
@@ -121,6 +123,39 @@ class Executor {
...
+        if (isFinite(buildTimeout) === false) {
+            throw new Error(`Invalid buildTimeout value: ${buildTimeout}`);
+        }
+
+        const options = {
+            uri: `${config.apiUri}/v4/builds/${config.buildId}/token`,
+            method: 'POST',
+            body: { buildTimeout },
+            headers: { Authorization: `Bearer ${config.token}` },
+            strictSSL: false,
```
Comment (jithin1987) [inline index.js]: why is this false ?

Comment (sakka2) [reply index.js]: It was wrong. changed true

File: .nyc_output/98b9c3d856e4e1493ac20e3f8e4c43a0.json [hunk 7]
```diff
@@ -0,0 +1 @@
...137,"column":40},"end":{"line":137,"column":76}},"type":"default-arg","locations":[{"start":{"line":137,"column":55},"end":{"line":137,"column":76}}],"line":137},"2":{"loc":{"start":{"line":138,"column":8},"end":{"line":140,"column":9}},"type":"if","locations":[{"start":{"line":138,"column":8},"end":{"line":140,"column":9}},{"start":{"line":138,"column":8},"end":{"line":140,"column":9}}],"line":138},"3":{"loc":{"start":{"line":153,"column":8},"end":{"line":155,"column":9}},"type":"if","locations":[{"start":{"line":153,"column":8},"end":{"line":155,"column":9}},{"start":{"line":153,"column":8},"end":{"line":155,"column":9}}],"line":153}},"s":{"0":15,"1":15,"2":15,"3":15,"4":15,"5":7,"6":7,"7":3,"8":4,"9":16,"10":2,"11":1,"12":1,"13":3,"14":2,"15":1,"16":1,"17":1,"18":1,"19":1,"20":2,"21":1,"22":1,"23":1,"24":4,"25":1,"26":3,"27":3,"28":3,"29":1,"30":2,"31":15},"f":{"0":7,"1":16,"2":2,"3":1,"4":1,"5":3,"6":2,"7":1,"8":1,"9":1,"10":1,"11":1,"12":2,"13":1,"14":1,"15":1,"16":4},"b":{"0":[3,4],"1":[1],"2":[1,3],"3":[1,2]},"_coverageSchema":"332fd63041d2c1bcb487cc26dd0d5f7d97098a6c","hash":"ec3193cc2c2211b08d53e513dda3cedcc848f43f","contentHash":"1875587c4753bfccb50eae5c2b432e9b_10.3.2"}}
```
Comment (jithin1987) [inline .nyc_output/98b9c3d856e4e1493ac20e3f8e4c43a0.json]: please remove this file

**Comment by sakka2** [comment]
sorry. I deleted.

**Pull request merged**